File: rnamake/motif_type.py
import exceptions


"""
Basic type information for motifs for looking up in databases and libraries

+-------------------+-------+---------------------------------------------+
|  Motif Type       | Value | Description                                 |
+===================+=======+=============================================+
|   TWOWAY          | 0     | Twoway junctions                            |
+-------------------+-------+---------------------------------------------+
|   NWAY            | 1     | Junctions with more then two ends           |
+-------------------+-------+---------------------------------------------+
|   HAIRPIN         | 2     | Motif with only one end                     |
+-------------------+-------+---------------------------------------------+
|   TCONTACT_HP_HP  | 3     | Tertiary contact between two hairpins,      |
|                   |       | two ends                                    |
+-------------------+-------+---------------------------------------------+
|   TCONTACT_H_HP   | 4     | Tertiary contact between one hairpin and    |
|                   |       | one helix, three ends                       |
+-------------------+-------+---------------------------------------------+
|   TCONTACT_H_H    | 5     | Tertiary contact between two helices,       |
|                   |       | four ends                                   |
+-------------------+-------+---------------------------------------------+
|   TWOWAY_SEGMENTS | 8     | Segments of multiple motifs with two ends   |
+-------------------+-------+---------------------------------------------+
|   HELIX           | 9     | RNA with only GC/AU pairs                   |
+-------------------+-------+---------------------------------------------+
|   SSTRAND         | 10    | Single stranded RNA                         |
+-------------------+-------+---------------------------------------------+
|   TCONTACT        | 11    | General tertiary contact                    |
+-------------------+-------+---------------------------------------------+
|   UNKNOWN         | 99    | Motif with unassigned motif type            |
+-------------------+-------+---------------------------------------------+
|   ALL             | 999   | Place holder type for iteration of motifs   |
+-------------------+-------+---------------------------------------------+

"""

TWOWAY            = 0
NWAY              = 1
HAIRPIN           = 2
TCONTACT_HP_HP    = 3
TCONTACT_H_HP     = 4
TCONTACT_H_H      = 5
TWOWAY_SEGMENTS   = 8
HELIX             = 9
SSTRAND           = 10
TCONTACT          = 11
UNKNOWN           = 99
ALL               = 999

type_to_str_dict = {
    TWOWAY            : 'TWOWAY',
    NWAY              : 'NWAY',
    HAIRPIN           : 'HAIRPIN',
    TCONTACT          : 'TCONTACT',
    TCONTACT_HP_HP    : 'TCONTACT_HP_HP',
    TCONTACT_H_HP     : 'TCONTACT_H_HP',
    TCONTACT_H_H      : 'TCONTACT_H_H',
    TWOWAY_SEGMENTS   : 'TWOWAY_SEGMENTS',
    HELIX             : 'HELIX',
    UNKNOWN           : 'UNKNOWN',
    ALL               : 'ALL'
}

str_to_type_dict = {
   'TWOWAY'           : TWOWAY,
   'NWAY'             : NWAY,
   'HAIRPIN'          : HAIRPIN,
   'TCONTACT_HP_HP'   : TCONTACT_HP_HP,
   'TCONTACT_H_HP'    : TCONTACT_H_HP,
   'TCONTACT_H_H'     : TCONTACT_H_H,
   'TWOWAY_SEGMENTS'  : TWOWAY_SEGMENTS,
   'HELIX'            : HELIX,
   'UNKNOWN'          : UNKNOWN,
   'ALL'              : ALL
}


def type_to_str(mtype):
    """
    convery motif_type enum value into a string for each printing

    :param mtype: motif type value
    :type mtype: motif_type
    :return: return string name of motif_type from type_to_str_dict
    :rtype: string

    :examples:

    ..  code-block:: python

        >>> import motif_type
        >>> motif_type.type_to_str(motif_type.TWOWAY)
        TWOWAY

    """

    if mtype in type_to_str_dict:
        return type_to_str_dict[mtype]
    else:
        raise exceptions.MotifTypeException("MotifType not recognized: " + mtype)


def str_to_type(type_name):
    """
    converts string to its corresponding motif_type

    :param type_name: name of motif_type enum
    :return: corresponding motif_type enum
    :rtype: motif_type

    ..  code-block:: python

        >>> import motif_type
        >>> motif_type.str_to_type('TWOWAY')
        0

    """

    if type_name in str_to_type_dict:
        return str_to_type_dict[type_name.upper()]
    else:
        raise exceptions.MotifTypeException("motif_type name not recognized")

File: src/util/motif_type.cc
//
//  motif_type.cpp
//  RNAMake
//
//  Created by Joseph Yesselman on 1/30/15.
//  Copyright (c) 2015 Joseph Yesselman. All rights reserved.
//

#include "util/motif_type.h"

namespace util {

String const
type_to_str(MotifType const mtype) {
    if (mtype == MotifType::TWOWAY) { return "TWOWAY"; }
    else if (mtype == MotifType::NWAY) { return "NWAY"; }
    else if (mtype == MotifType::HAIRPIN) { return "HAIRPIN"; }
    else if (mtype == MotifType::T_T) { return "2X_TWOWAY"; }
    else if (mtype == MotifType::T_T_T) { return "3X_TWOWAY"; }
    else if (mtype == MotifType::TWOWAY_SEGMENTS) { return "TWOWAY_SEGMENTS"; }
    else if (mtype == MotifType::HELIX) { return "HELIX"; }
    else if (mtype == MotifType::UNKNOWN) { return "UNKNOWN"; }
    else if (mtype == MotifType::TCONTACT) { return "TCONTACT"; } // Added by CJ 08/20
    else { throw "cannot indentify type "; }
}

MotifType const
str_to_type(String const s) {
    if (s.compare("TWOWAY") == 0) { return MotifType::TWOWAY; }
    else if (s.compare("NWAY") == 0) { return MotifType::NWAY; }
    else if (s.compare("HAIRPIN") == 0) { return MotifType::HAIRPIN; }
    else if (s.compare("2X_TWOWAY") == 0) { return MotifType::T_T; }
    else if (s.compare("3X_TWOWAY") == 0) { return MotifType::T_T_T; }
    else if (s.compare("TWOWAY_SEGMENTS") == 0) { return MotifType::TWOWAY_SEGMENTS; }
    else if (s.compare("HELIX") == 0) { return MotifType::HELIX; }
    else if (s.compare("TCONTACT") == 0) { return MotifType::TCONTACT; } //added by CJ 08/20
    else if (s.compare("UNKNOWN") == 0) { return MotifType::UNKNOWN; }
    else { throw "cannot indentify str for type"; }

}

}



